ZIP-99 Update Python and dependencies

rute-santos opened the pull request:
Python was updated to 3.12.
All dependencies were updated (hopefully).
This cause an increase in the size of the downloader and on-demand functions, which was fixed by always uploading the code to S3 and updating the function using the S3 content.

rute-santos committed:
Update to python 3.12 and upgrade all dependencies.

Pin version of rsa==4.7.2 to fix the issue where it was installing a really old version of awscli.
This was due to awscli requirements: rsa>=3.1.2,<4.8 while latest rsa was installed (4.9.1).

rute-santos committed:
Always use S3 when updating function code due to their size.

rute-santos committed:
Fix lint/format errors.

lbjay commented on requirements/tox.txt:
```diff
@@ -1,166 +1,161 @@
+pyparsing==3.2.3
     # via
-    #   -r requirements/../function_requirements/common-requirements.txt
+    #   -r /Users/ras300/Documents/opencast-16.x/zoom-ingester-oc-rute-16x/function_requirements/common-requirements.txt
```
These are just annotation comments but it looks strange. I wonder why it's doing this.

lbjay commented on functions/utils/status.py:
```diff
@@ -154,9 +154,9 @@ def set_pipeline_status(
-            expression_attribute_values[
-                ":recording_processing"
-            ] = ZoomStatus.RECORDING_PROCESSING.name
+            expression_attribute_values[":recording_processing"] = (
```
This looks odd. Was it a formatting change made by black?

lbjay reviewed (approved):
I think this is good. Just had one question about the requirements files annotations.

lbjay committed:
initial stab at packaging functions and their dependencies as a docker image

Major changes:
- removed all packaging/zip/upload functionality from `tasks.py`
- consolidated functions and requirements
- introduced a Dockerfile for packaging into a docker image
- updated the CDK lambda function construct to build and use an image
  asset
- removed codebuild components

rute-santos committed:
Make sure all functions have new names to facilitate deployment to existing stacks.

rute-santos committed:
Add platform to docker build.

rute-santos committed:
Merge pull request #172 from harvard-dce/docker-runtime

Docker runtime

lbjay committed:
readme updates
